### DISCUSSION TITLE
Due to the large size of the JSON in Mockoon, it leads to slow API response.

### DISCUSSION THREAD

**Discussion by heyyuyu** [mockoon/mockoon#1407] [General]
Hi, I generate Mockoon JSON files automatically. The API response becomes slow when the JSON file size is too large. I think this is a disadvantage when using JSON configuration. Can this be optimized?Maybe it would be better to split the JSON into smaller files based on different modules and package them as assets to execute; this might be more efficient.

**Comment by 255kb** [comment]
Hi, I'm not sure to understand which part is slow. 
Is the UI unresponsive? What do you mean by "generating JSON files"? Do you mean generating a response using templating takes time? 

Thanks.

**Reply by heyyuyu** [reply]
Sorry,I found its not slow but the response failed when request frequently.
the error shows below:
org.apache.http.NoHttpResponseException: 7.189.117.126:4026 failed to respond
![image](https://github.com/mockoon/mockoon/assets/19744761/318fe679-581c-4196-81b4-d3ecad8e4053)

**Reply by 255kb** [reply]
There are a lot of things that could go wrong, and without more information I won't be able to help. 
Could you share your environment data file, so I can try to reproduce? 

[Finding your environment data file](https://mockoon.com/docs/latest/mockoon-data-files/data-storage-location/#locating-the-files)

You can upload it here or send it to support@mockoon.com too. 

If you are not able to share the data file, could you at least provide an example of the template you are trying to generate? 
Thanks

**Reply by heyyuyu** [reply]
I can't upload due to the network limit; this error is displayed in the Mockoon CLI. This is an occasional issue. The performance test can reproduce this problem. 
Here is the JMeter configuration:
10 threads continuously requesting the same URL.
And request is : 

> GET https://7.189.117.126:4026/rest/openapi/server/device/detail?dn=254343
Connection: keep-alive
Host: 7.189.117.126:4026
User-Agent: Apache-HttpClient/4.5.10 (Java/1.8.0_252)

Response body is:

> {"code":0,"data":[{"dn":"254343","parentDn":"/","ipAddress":"172.21.16.57","name":"test004","type":"RH2288H V3","uuid":"FB24EEBE-BED1-11E6-A0EC-4CF95DAD695D","status":"0","macAddr":"4c:f9:5d:ad:69:5d","bmcMacAddr":"4c:f9:5d:ad:69:5d","averagePower":"238","peakPower":"450","powerConsumption":"4795.060","powerStatisticStartTime":"2014-01-01 00:00:58","serverSource":"","serverFlag":"","serverState":"","dnsSource":"Static","dnsName":"","dnsServerIP":"","location":"","bmaVersion":"","bmaDriverVersion":"","osVersion":"","osKernelVersion":"","hostName":"","domainName":"","hostDesc":"","cpuNums":"0","cpuCores":"0","MemoryCapacity":"0 MB","CPU":[{"name":"CPU_743_66422_254343_1","healthState":"0","frequency":"2300 MHz","manufacture":"Intel(R) Corporation","model":"Intel(R) Xeon(R) CPU E5-2650 v3 @ 2.30GHz","moId":"129","uuid":"CPU_743_66422_254343_1","presentState":"1","enableState":"","processorId":"","totalCores":"","totalThreads":"","l1CacheKiB":"","l2CacheKiB":"","l3CacheKiB":"","partNumber":""}],"Memory":[{"name":"Memory_743_66422_254343_2","healthState":"0","capacity":"32768","manufacture":"Samsung","frequency":"2400","moId":"714","uuid":"Memory_743_66422_254343_2","presentState":"1","memoryType":"DDR4","serialNumber":"1920150F","minVoltage":"1200","rankCount":"2","dataWidthBits":"72","technology":"Synchronous| Registered (Buffered)","partNumber":"03024BFQ"}],"Disk":[],"PSU":[],"Fan":[],"board":[],"RAID":[],"NetworkCard":[],"PCIE":[],"NetCard":[],"GPU":[],"DiskPartition":[],"Mezz":[]}],"description":"Succeeded in querying server details."}

Response header is:

> HTTP/1.1 200 OK
> Content-Type: application/json; charset=utf-8
> Content-Length: 1561
> Date: Thu, 06 Jun 2024 13:34:56 GMT
> Connection: keep-alive
> Keep-Alive: timeout=5

**Reply by 255kb** [reply]
Are you doing load testing with mockoon? 
Mockoon was never built for load testing in the first place. It should be performant enough, but if 10 threads are making a lot of requests, it's not a use case we want to cover.

**Reply by heyyuyu** [reply]
Due to server get the same error when request mockoon client frequently.So I reproduce this problem with JMeter.

**Reply by 255kb** [reply]
ok, but what do you mean by frequently ? 
We have instances of mockoon running in the cloud with very frequent requests and they are fine, not consuming too much ram or CPU, responding in time, etc. You are not even using templating (which can be heavy if you generate huge templates), so without more info I can't do anything. Sorry.

**Reply by 255kb** [reply]
The only issue I see is that you cannot upload your environment file du to size limit, which let me think that it's huge, probably too big for the application. 
Splitting wouldn't help because anyway the whole configuration is needed to run the server properly with all your setup.

**Reply by heyyuyu** [reply]
Yes! It's a huge file because of over 4000 request rules.Thank you anyway.

**Reply by 255kb** [reply]
Then it's probably too heavy for our system if everytime there is a request it evaluates 4000 rules. 

Splitting the data or not wouldn't change much I guess. 
What I would suggest is to create only one response, and use templating to generate a different body response depending on the query parameters, with some random data. This would be way faster. But if you are interested in having real data, it may not solve your issue.
